libs/vkd3d/command.c

```diff
@@ -3298,31 +3328,43 @@ static void d3d12_command_list_update_root_descriptors(struct d3d12_command_list
                 list->allocator, root_signature->vk_root_descriptor_layout, VKD3D_DESCRIPTOR_POOL_TYPE_STATIC);
     }
 
-    /* TODO bind null descriptors for inactive root descriptors */
-    bindings->root_descriptor_dirty_mask &= bindings->root_descriptor_active_mask;
-
-    while (bindings->root_descriptor_dirty_mask)
+    if (!(list->device->bindless_state.flags & VKD3D_RAW_VA_ROOT_DESCRIPTOR))
     {
-        root_parameter_index = vkd3d_bitmask_iter64(&bindings->root_descriptor_dirty_mask);
-        root_parameter = root_signature_get_root_descriptor(root_signature, root_parameter_index);
+        /* TODO bind null descriptors for inactive root descriptors */
+        bindings->root_descriptor_dirty_mask &= bindings->root_descriptor_active_mask;
 
-        if (!vk_write_descriptor_set_from_root_descriptor(list,
-                &descriptor_writes[descriptor_write_count], root_parameter,
-                descriptor_set, &bindings->root_descriptors[root_parameter_index]))
-            continue;
+        while (bindings->root_descriptor_dirty_mask)
+        {
+            root_parameter_index = vkd3d_bitmask_iter64(&bindings->root_descriptor_dirty_mask);
+            root_parameter = root_signature_get_root_descriptor(root_signature, root_parameter_index);
 
-        descriptor_write_count += 1;
+            if (!vk_write_descriptor_set_from_root_descriptor(list,
+                    &descriptor_writes[descriptor_write_count], root_parameter,
+                    descriptor_set, &bindings->root_descriptors[root_parameter_index]))
+                continue;
+
+            descriptor_write_count += 1;
+        }
     }
 
     if (root_signature->flags & VKD3D_ROOT_SIGNATURE_USE_INLINE_UNIFORM_BLOCK)
     {
-        d3d12_command_list_fetch_inline_uniform_block_data(list, bind_point, inline_uniform_block_data);
+        d3d12_command_list_fetch_inline_uniform_block_data(list, bind_point, &root_parameter_data);
 
         vk_write_descriptor_set_and_inline_uniform_block(&descriptor_writes[descriptor_write_count],
-                &inline_uniform_block_write, descriptor_set, root_signature, inline_uniform_block_data);
+                &inline_uniform_block_write, descriptor_set, root_signature, &root_parameter_data);
 
         descriptor_write_count += 1;
     }
+    else if (bindings->root_descriptor_dirty_mask)
+    {
+        va_count = d3d12_command_list_fetch_root_descriptor_vas(list, bind_point, &root_parameter_data);
+
+        VK_CALL(vkCmdPushConstants(list->vk_command_buffer,
```

Comment:
Do push constants get pushed with an offset in this case? I didn't see any change for that.

Reply:
`d3d12_root_signature_init_push_constants` takes care of the offset:
```c
        root_signature->parameters[i].constant.constant_index = push_constant_range->size / sizeof(uint32_t);
        root_signature->parameters[i].constant.constant_count = p->Constants.Num32BitValues;
```
where `push_constant_range->size` will be the total size of the VA portion for the first set of constants.

Reply:
Right, looked at the three relevant vkCmdPushConstants, and offsets were already applied.